### DISCUSSION TITLE
getting HTML element data attributes in DOMCrawler

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#59242] [Q&A] [answered]
I know it's possible to get a single data attribute from the DOM Crawler with
```php
                $blocks->each(function (Crawler $node, $i) use ($componentHtml) {
                    $el = $node->attr('data-element');
                    // $elData = $node->data(); // this doesn't work.
}
```
but is there a way to get an array of all the data attributes, like javascripts el.dataset?  I guess I could iterate over all the children and extract the ones beginning with data-, but I'm wondering if there's already a tool to do that.  Or maybe DOMCrawler isn't the right component to use?

**Comment by MatTheCat** [comment]
Iterating over attributes is trivial so I’m not sure you’d need a dedicated tool. I guess you could try adding the crawler this functionality in a PR.

**Reply by tacman** [reply]
how do I iterative over all the attributes in the DOM crawler?  I could then look for 'data-' and populate an array.  I'm sure it's trivial, but I can't seem to find it.   I still find domNode and domElement confusing.

**Reply by MatTheCat** [reply, marked as answer]
The crawler exposes a `getNode` method returning a `DOMNode`, which exposes an `$attributes` property yielding `DomAttr`s:

```php
foreach ($crawler->getNode(0)->attributes as $attribute) {
    dump($attribute->name, $attribute->value);
}
```

**Reply by tacman** [reply]
Thanks!
